include/kernel.h

```diff
@@ -3357,8 +3357,15 @@ struct k_pipe {
  *
  * @return N/A
  */
-__syscall void k_pipe_init(struct k_pipe *pipe, unsigned char *buffer,
-			   size_t size);
+void k_pipe_init(struct k_pipe *pipe, unsigned char *buffer, size_t size);
+
+/**
+ * @brief Initialize a pipe and allocate a buffer for it
+ *
+ * Storage for the buffer region will be allocated from the calling thread's
+ * resource pool. This memory will not ever be released.
```

Comment:
I understand why the memory for this object is never released, but this makes me uneasy.  There might be some gotchas (e.g. a k_pipe allocated this way being used by a supervisor thread, and the user thread that allocates this dies), but I would take a look at making these allocations reference-counted.

Maybe if you repurpose the permission bitmap and use __builtin_popcount() as a refcount you could get away with most of the boilerplate?  (If popcount is 0 on thread exit, k_free() can be called.)

Reply:
I think I see where you are going with this.

> Maybe if you repurpose the permission bitmap 

The problem is that supervisor threads can do anything with any object they want. They do not need to be assigned permission. We could be in the middle of an API call from a supervisor thread and suddenly free the buffer if all we were tracking were permission settings. Maybe we enforce that supervisor mode has to self-grant permissions or undefined behavior may take place? I think I'm OK with that.

> (If popcount is 0 on thread exit, k_free() can be called.)

I see where to do this. We could check it in _thread_perms_all_clear, specifically in the clear_perms_cb(). It would make this a little slower but I don't think that's a big deal.

Thinking some more about this...

Reply:
Worth noting: currently, this mechanism is orthogonal to user mode.

If we add reference counting / automatic freeing of resources, we would have to either:
- Document that resource releasing only happens if CONFIG_USERSPACE is enabled, but otherwise leave as-is
- #ifdef out all the new APIs which do behind the scenes allocation

I think I prefer the first option.

Reply:
I can think of situations where this can blow up.

Say we have threads A and B. They mutually have permissions on each other and some k_pipe we'll call P.

A enters an API call to do something with P. A is preempted while in the syscall.
B is scheduled. B calls k_object_access_revoke(P, B) and k_object_access_revoke(P, A)
P no longer has any references, so the final call to k_object_access_revoke() initiates logic to free P's memory buffer.
A swaps back in to continue processing the system call and promptly explodes while in supervisor mode.
Hmmm.

Reply:
Maybe, to supplement the permission/reference bitmap, we additionally increment/decrement some per-object counter upon API entry/exit. That sounds horrible to maintain.

Could also make k_object_access_revoke() supervisor-only.